### PR TITLE
Support sending streams for head blob announce

### PR DESCRIPTION
**Pull request by kaykurokawa**
Enables sending of streams for Head blob announce
depends on https://github.com/lbryio/reflector-cluster/pull/5 and https://github.com/lbryio/lbry/pull/872

This will switch prism to send out streams to hosts instead of blobs.

We create class PrismStreamClient, coded in pretty the same way as PrismClient (for sending blobs). 
PrismStreamClientFactory is the corresponding factory, and we introduce function build_prism_stream_client_factory() that will do all the prep work to make sure that we can actually send the stream (check if we have all the blobs). We introduce a similar function build_prism_blob_client_factory() for building PrismClientFactory (for sending blobs). 

Functional tests have been split into two files for testing PrismStreamClient and PrismClient. To support testing, we now have an additional argument in process_blob() that is a deferred used to initialize the test.

### CODE REVIEW

**Comment by kaykurokawa** [comment]
added two commits 

a61fcdefc34443406452cccfc89c23d23cc3574c refactors _process_blob() and _process_stream() into one function,  update_sent_blob() is also refactored to update_sent_blobs() to loop through multiple blobs as suggested in https://github.com/lbryio/reflector-cluster/pull/5

67be3dc3e5214882f16b2b03572ff98669750822 errors back if connectionLost is called with reason other than ConnectionDone, in this case we log the reason, try to update any blobs that might have gotten sent , and close reactor, as suggested in https://github.com/lbryio/reflector-cluster/pull/5

**Comment by kaykurokawa** [comment]
Fixed enqueue_blob() to be called properly (forgot to use updated arguments)

File: prism/storage/storage.py [hunk 1]
```diff
@@ -13,45 +13,62 @@
...
-        return threads.deferToThread(self.db.sadd, name, *values)
+        return self.defer_func(self.db.sadd, name, *values)
 
+    def sdiff(self, name, *values):
+        return self.defer_func(self.db.sdiff, name, *values)
 
 class ClusterStorage(object):
-    def __init__(self, path=None):
-        self.db = RedisHelper()
+    def __init__(self, path=None, redis_address=conf['redis server']):
```
Comment (jackrobison) [inline prism/storage/storage.py]: I think redis_address should default to None, and then `self.db = RedisHelper(redis_address or conf['redis server'])`

File: tests/functional/test_blob_task.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+import multiprocessing
+import threading
+import Queue
+
+# this turns on logging
+from twisted.python import log
+import sys
+log.startLogging(sys.stdout)
+
+sys.path.insert(0, os.path.dirname(__file__))
```
Comment (jackrobison) [inline tests/functional/test_blob_task.py]: is this necessary?

File: prism/protocol/task.py [hunk 3]
```diff
@@ -48,36 +48,95 @@ def next_host():
...
+    d.addCallback(lambda factory: connect_factory(host, port, factory, blob_storage, sd_hash))
+    reactor.run()
+    return sys.exit(0)
+
+
 @retry_redis
-def enqueue_blob(blob_hash, client_factory_class):
-    q = Queue(connection=redis_conn)
-    q.enqueue(process_blob, blob_hash, client_factory_class, timeout=60)
+def enqueue_stream(sd_hash, num_blobs_in_stream, db_dir, client_factory_class, redis_address=settings['redis server'],
```
Comment (jackrobison) [inline prism/protocol/task.py]: Same nit about dicts for default values instead of None

File: tests/functional/test_blob_task.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        client_factory_class = build_prism_blob_client_factory
+
+        # start client
+        from twisted.internet import reactor
+        reactor.addSystemEventTrigger('before','shutdown', self._on_finish_blob)
+        try:
+            process_blob(BLOB_HASH, self.client_db_dir, client_factory_class, 'fake', host_infos=('localhost',5566,0),
+                        setup_d = self._setup_client_blobs)
+        except SystemExit:
+            pass
```
Comment (jackrobison) [inline tests/functional/test_blob_task.py]: Maybe a test where forwarding the blob fails / sys.exit(1) is called?

File: prism/storage/storage.py (hunk 1 above)
Comment (kaykurokawa) [reply prism/storage/storage.py]: changed here b68da3d in https://github.com/lbryio/reflector-cluster/pull/8

File: tests/functional/test_blob_task.py (hunk 2 above)
Comment (kaykurokawa) [reply tests/functional/test_blob_task.py]: couldn't get travis to work without this..

**Comment by kaykurokawa** [comment]
closing this in favor of https://github.com/lbryio/reflector-cluster/pull/8

**Pull request closed without merging**